src/backend/farmer/node_client.rs

```diff
@@ -0,0 +1,693 @@
+};
+use sc_consensus_subspace::notification::SubspaceNotificationStream;
+use sc_consensus_subspace::slot_worker::{NewSlotNotification, RewardSigningNotification};
+use sc_rpc::SubscriptionTaskExecutor;
+use sc_utils::mpsc::TracingUnboundedSender;
+use sp_api::{ApiError, ProvideRuntimeApi};
+use sp_consensus::SyncOracle;
+use sp_consensus_subspace::{ChainConstants, FarmerPublicKey, FarmerSignature, SubspaceApi};
+use sp_core::{ByteArray, H256};
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::num::NonZeroUsize;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::{Arc, Weak};
+use std::time::Duration;
+use subspace_archiving::archiver::NewArchivedSegment;
+use subspace_core_primitives::crypto::kzg::Kzg;
+use subspace_core_primitives::{
+    BlockHash, HistorySize, PieceIndex, PublicKey, SegmentHeader, SegmentIndex, SlotNumber,
+    Solution,
+};
+use subspace_fake_runtime_api::RuntimeApi;
+use subspace_farmer::node_client::{Error, NodeClientExt};
+use subspace_farmer::NodeClient as NodeClientT;
+use subspace_farmer_components::FarmerProtocolInfo;
+use subspace_networking::libp2p::Multiaddr;
+use subspace_rpc_primitives::{
+    FarmerAppInfo, RewardSignatureResponse, RewardSigningInfo, SlotInfo, SolutionResponse,
+};
+use subspace_service::FullClient;
+use tracing::{debug, error, warn};
+
+/// This is essentially equal to expected number of votes per block, one more is added implicitly by
+/// the fact that channel sender exists
+const SOLUTION_SENDER_CHANNEL_CAPACITY: usize = 9;
+const REWARD_SIGNING_TIMEOUT: Duration = Duration::from_millis(500);
+
+/// Configuration for `NodeClient`
+pub(in super::super) struct NodeClientConfig {
+    /// Substrate client
+    pub client: Arc<FullClient<RuntimeApi>>,
+    /// Task executor that is being used by RPC subscriptions
+    pub subscription_executor: SubscriptionTaskExecutor,
+    /// New slot notification stream
+    pub new_slot_notification_stream: SubspaceNotificationStream<NewSlotNotification>,
+    /// Reward signing notification stream
+    pub reward_signing_notification_stream: SubspaceNotificationStream<RewardSigningNotification>,
+    /// Archived segment notification stream
+    pub archived_segment_notification_stream:
+        SubspaceNotificationStream<ArchivedSegmentNotification>,
+    /// DSN bootstrap nodes
+    pub dsn_bootstrap_nodes: Vec<Multiaddr>,
+    /// Segment headers store
+    pub segment_headers_store: SegmentHeadersStore<FullClient<RuntimeApi>>,
+    /// Subspace sync oracle
+    pub sync_oracle: Arc<dyn SyncOracle + Send + Sync>,
+    /// Kzg instance
+    pub kzg: Kzg,
+}
+
+#[derive(Default)]
+struct BlockSignatureSenders {
+    current_hash: H256,
+    senders: Vec<async_oneshot::Sender<RewardSignatureResponse>>,
+}
+
+/// In-memory cache of last archived segment, such that when request comes back right after
```

Comment:
This should only be necessary for RPC server. For in-process client you should be able to access pieces directly. Might be okay for now, but eventually this is definitely suboptimal.

We might want to change the API for example such that `subscribe_archived_segment_headers` gives you a `Box<dyn Something>` that you can extract segment header from, but then also iterate over all pieces in that segment. And on drop it'll do the acknowledgemnet automatically. Then we'll be able to more or less return a thin wrapped around `ArchivedSegmentNotification` instead of doing all of the tracking and matching that RPC server has to do. While RPC client can still wrap the same logic it does today behind nicer API,  have alrady done this with `Farm` and other traits on farmer.

I understand this might seem outside of the scope, but this is the right way of thinking about the problem. Code is not static, we can change and improve APIs when they don't seem to match what we're trying to do particularly well.